### DISCUSSION TITLE
Keys 🔖

### DISCUSSION THREAD

**Discussion by mediumhust** [TinyHust/notes#249] [Ideas]
- [ ] kubernetes websocket ingress
  - https://github.com/iamdejan/socketio-chat-app
  - https://github.com/sw360cab/websockets-scaling
- [ ] [HTTP authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#restricting_access_with_nginx_and_basic_authentication)
- [ ]  [Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)
- [ ] [scrollIntoView](https://www.w3schools.com/jsref/met_element_scrollintoview.asp)

**Comment by mediumhust** [comment]
## CSS
- webkit-box-reflect
- [mask-image](https://www.w3schools.com/cssref/css3_pr_mask-image.php)
- mix-blend-mode

**Comment by mediumhust** [comment]
## Algorithms

- Rolling, Canary, Blue-Green Deployment
- [SaaS Multi-Tenant Strategies | Silo vs Bridge vs Pool](https://d1.awsstatic.com/whitepapers/saas-tenant-isolation-strategies.pdf)
  - [SaaS Boost](https://aws.amazon.com/partners/saas-boost/)
    - [aws-saas-boost](https://github.com/awslabs/aws-saas-boost/blob/main/docs/getting-started.md#step-1---set-up-tooling) 
- Mutex and Semaphore
  - [Implementing Basic Lock Mechanisms for Parallel Processing in Node.js with Redis](https://medium.com/@dev.yusufbas/implementing-basic-lock-mechanisms-for-parallel-processing-in-node-js-with-redis-b4e2ee9d42e7)
  - [redis-semaphore](https://github.com/swarthy/redis-semaphore)

**Comment by mediumhust** [comment]
## SQL
- OLTP (On-line transactional processing)
- OLAP (On-line analytical processing)

**Reply by mediumhust** [reply]
MVCC (Multiversion concurrency control)

**Reply by mediumhust** [reply]
PostgreSQL Vacuum

**Reply by mediumhust** [reply]
- index
- partition

**Comment by mediumhust** [comment]
### TS
- decorator
  - https://www.typescriptlang.org/docs/handbook/decorators.html
  - https://mirone.me/a-complete-guide-to-typescript-decorator/

**Comment by mediumhust** [comment]
graphql introspection

**Comment by mediumhust** [comment]
### JS
- https://github.com/immerjs/immer
- https://github.com/tj/commander.js
  - https://nest-commander.jaymcdoniel.dev/en/introduction/intro/
- https://github.com/SBoudrias/Inquirer.js

**Reply by mediumhust** [reply]
- [method spoofing](https://expressjs.com/en/resources/middleware/method-override.html)

**Comment by mediumhust** [comment]
### Icons
- https://iconify.design/
- https://simpleicons.org/

**Comment by mediumhust** [comment]
Imperative vs Declarative

**Comment by mediumhust** [comment]
## Web
- [Content Security Policy (CSP)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)
- [scrollIntoView](https://www.w3schools.com/jsref/met_element_scrollintoview.asp)

**Comment by mediumhust** [comment]
### Data Engineering
- ETL (Extract, Transform, Load)
- CDC (Change Data Capture)
- SCD (Slowly changing dimension)

**Comment by mediumhust** [comment]
## Network
- Wake-on-LAN (WOL)
- Network Interface Card (NIC)

**Reply by mediumhust** [reply]
MITM (Man-in-the-Middle)
- [mitmproxy](https://mitmproxy.org/)
- [mitm](http://mitm.it/)
- [apk-mitm](https://github.com/niklashigi/apk-mitm)

**Comment by mediumhust** [comment]
## Virtualization
- Bare Metal
- Virtual Machine
---
- OS-level virtualization
- Full Virtualization
  - Hypervisor Type 1: 
    - Hyper-V 
    - Xen 
      - Citrix Hypervisor
      - XCP-ng
    - vSphere
      - ESXi hypervisor
    - KVM
      - Proxmox
  - Hypervisor Type 2: VirtualBox, Parallels Desktop, VMware Workstation,…
---
Nested Virtualization

**Comment by mediumhust** [comment]
## Bash Scripting
- heredoc
- [shell parameter expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html)
- [shebang](https://en.wikipedia.org/wiki/Shebang_%28Unix%29)

**Comment by mediumhust** [comment]
## OTT
- XMPP
- OMEMO
- FreeSWITCH

**Comment by mediumhust** [comment]
### Finance
- [plaid](https://plaid.com/)
- [dwolla](https://www.dwolla.com/)
- [alpaca](https://alpaca.markets/)

**Comment by mediumhust** [comment]
## AI

### Agent
- [langgraph](https://langchain-ai.github.io/langgraph/)
- [crewai](https://www.crewai.com/open-source)
- [autogen](https://github.com/microsoft/autogen)
---
- [MCP](https://github.com/modelcontextprotocol/servers)
- [A2A](https://github.com/google/A2A)
- Variational Autoencoder (VAE)
- Low-rank adaptation (LoRA)

**Comment by mediumhust** [comment]
## Database
- join:
  - hash join
- index:
  - bitmap index
- Data Manipulation Language (DML)
- shared storage (SAN)

### MongoDB
- Storage Engine
- executionStats
  - `db.customers.find().explain("executionStats")`
- [Replication lag ](https://docs.aws.amazon.com/documentdb/latest/developerguide/replication.html)

### Postgresql
- execution plan / plan cache
- Multi-Version Concurrency Control (MVCC)
- VACUUM — garbage-collect and optionally analyze a database
- [PL/pgSQL](https://neon.tech/postgresql/postgresql-plpgsql)
- [patroni](https://github.com/patroni/patroni)
- connection pool
  - [Pgpool](https://www.pgpool.net/mediawiki/index.php/Main_Page)
  - [PgBouncer](https://www.pgbouncer.org/)
- [temporary table](https://neon.tech/postgresql/postgresql-tutorial/postgresql-temporary-table)

### Oracle SQL
- Real Application Clusters (RAC)
- GoldenGate CDC
- Data Guard
- Archive log
- Redo log

### Sql server
- high-availability and disaster-recovery
  - log shipping
  - failover cluster instances (FCI)
  - database mirroring
  - always on availability groups (AG)
  - replication

**Comment by mediumhust** [comment]
## Computer science
- NLP (Natural Language Processing)
  - Tokenization
    -  Word Tokenizer
    - Sentence Tokenizer
  - Stemming
  - Lemmatization
  - stopwords 
  - Bag Of Words (BoW): a representation of text that describes the occurrence of words within a document
  - Collocations
  - Word Embedding
    - word2vec (Google)
    - Glove (Stanford)
    - fastest (Facebook)
  - RNN ( Recurrent Neural Network ) 
  - CNN ( Convolutional Neural Network ) 
  - IVR ( Interactive Voice Response )

**Reply by mediumhust** [reply]
### Machine Learning
- K-means [1](https://github.com/AarohiSingla/K-Means/blob/master/K-Means%20-%20Copy.ipynb) [2](https://machinelearningcoban.com/2017/01/01/kmeans/)
  - Elbow Method
  - Silhouette Score
- KNN ( K-nearest neighbors ) [1](https://github.com/AarohiSingla/Knn-Algorithm/blob/master/fruit_dataset_solution.ipynb) [2](https://machinelearningcoban.com/2017/01/08/knn/)
- Linear Regression
- Gradient Descent (GD)
  - Batch Gradient Descent
  - Stochastic Gradient Descent (SGD)
  - Mini-batch Gradient Descent (mini-batch GD)
- Perceptron Learning Algorithm (PLA)
  - Pocket Algorithm
- Logistic Regression
- Binary Classifiers
  - one-vs-rest (ovr)
- Softmax Regression
- Multi-layer Perceptron (MLP)
- terms:
  - level sets
  - learning rate: $`\eta`$
  - epoch
  - Stopping Criteria
  - saddle points
  - Neural Networks
    - summation function
    - activation function
      - sigmoid $`f(s) = \frac{1}{1 + e^{-s}} \triangleq \sigma(s)`$
      - tanh: $`\text{tanh}(s) = \frac{e^{s} - e^{-s}}{e^s + e^{-s}}`$
      - ReLU (Rectified Linear Unit) $`f(s) = \max(0, s)`$
    - Supervised
    - Unsupervised
      - Generative Adversarial Network (GAN)
      - Deep Convolutional Generative Adversarial Network (DCGAN)
  - projection matrix
  - Feature Scaling and Normalization
    - Rescaling: $`x’ = \frac{x - \min(x)}{\max(x) - \min(x)}`$
    - Standardization: $`x’ = \frac{x - \bar{x}}{\sigma}`$
    - Scaling to unit length: $`\mathbf{x}’ = \frac{\mathbf{x}}{||\mathbf{x}||_2}`$
  - element-wise product or Hadamard product $`\odot`$
  - the equation of any tangent at $`x_0`$ is: $`g(x)=f'(x_0)(x-x_0)+f(x_0)`$
  - [polynomial interpolation Lagrange](https://vuontoanblog.blogspot.com/2012/10/polynomial-interpolation-lagrange.html): $`P(x) = \sum_{i=1}^{n+1} y_i \prod_{j \neq i}\frac{x-x_j}{x_i-x_j}`$
  - [Cross Entropy](https://machinelearningcoban.com/2017/02/17/softmax/#-cross-entropy)
  - Loss function
    - MAE (Mean absolute error)

**Reply by mediumhust** [reply]
## Computer vision
- HOG - Histrogram of oriented gradient

**Comment by mediumhust** [comment]
## Tech stacks
- https://zerodha.tech/blog/1-5-million-pdfs-in-25-minutes/
  - [typst](https://github.com/typst/typst)
  - [nomad](https://github.com/hashicorp/nomad)
  - [rundeck](https://www.rundeck.com/)
  - [haraka](https://github.com/haraka/Haraka)
  - [dungbeetle](https://github.com/zerodha/dungbeetle)
- [logging](https://zerodha.tech/blog/logging-at-zerodha/)
  - [clickhouse](https://clickhouse.com/)
  - [vector](https://vector.dev/)
  - [metabase](https://www.metabase.com/)

**Comment by mediumhust** [comment]
## Probability
- kernel density estimates (KDEs)
- empirical cumulative distribution functions (ECDFs)
- Root Mean Squared Error (RMSE)
- probability mass function (PMF)
- probability density function (PDF)
- cumulative distribution function (CDF)

**Comment by mediumhust** [comment]
## DataOps
- data lakehouse open source
- open lakehouse table format (OLTF)
